src/workerd/api/gpu/gpu-buffer.c++

```diff
@@ -3,61 +3,181 @@
 //     https://opensource.org/licenses/Apache-2.0
 
 #include "gpu-buffer.h"
-#include <_types/_uint64_t.h>
+#include "workerd/io/io-context.h"
+#include "workerd/jsg/exception.h"
+#include "workerd/jsg/jsg.h"
 
 namespace workerd::api::gpu {
 
-GPUBuffer::GPUBuffer(wgpu::Buffer b, wgpu::BufferDescriptor desc)
-    : buffer_(kj::mv(b)), desc_(kj::mv(desc)) {
+GPUBuffer::GPUBuffer(jsg::Lock& js, wgpu::Buffer b, wgpu::BufferDescriptor desc,
+                     wgpu::Device device, kj::Own<AsyncRunner> async)
+    : buffer_(kj::mv(b)), device_(kj::mv(device)), desc_(kj::mv(desc)),
+      async_(kj::mv(async)), detachKey_(new jsg::V8Ref<v8::Object>(
+                                 js.v8Isolate, v8::Object::New(js.v8Isolate))) {
 
   if (desc.mappedAtCreation) {
     state_ = State::MappedAtCreation;
   }
 };
 
-kj::ArrayPtr<kj::byte>
-GPUBuffer::getMappedRange(jsg::Optional<GPUSize64> offset,
+v8::Local<v8::ArrayBuffer>
+GPUBuffer::getMappedRange(jsg::Lock& js, jsg::Optional<GPUSize64> offset,
                           jsg::Optional<GPUSize64> size) {
 
   JSG_REQUIRE(state_ == State::Mapped || state_ == State::MappedAtCreation,
               TypeError, "trying to get mapped range of unmapped buffer");
 
-  uint64_t o = 0;
-  KJ_IF_MAYBE (real_offset, offset) {
-    o = *real_offset;
-  }
-
-  uint64_t s = desc_.size - o;
-  KJ_IF_MAYBE (real_size, size) {
-    s = *real_size;
-  }
+  uint64_t o = offset.orDefault(0);
+  uint64_t s = size.orDefault(desc_.size - o);
 
   uint64_t start = o;
   uint64_t end = o + s;
-  for (auto &mapping : mapped_) {
+  for (auto& mapping : mapped_) {
     if (mapping.Intersects(start, end)) {
-      KJ_FAIL_REQUIRE("mapping intersects with existing one");
+      JSG_FAIL_REQUIRE(TypeError, "mapping intersects with existing one");
     }
   }
 
-  auto *ptr = (desc_.usage & wgpu::BufferUsage::MapWrite)
+  auto* ptr = (desc_.usage & wgpu::BufferUsage::MapWrite)
                   ? buffer_.GetMappedRange(o, s)
-                  : const_cast<void *>(buffer_.GetConstMappedRange(o, s));
+                  : const_cast<void*>(buffer_.GetConstMappedRange(o, s));
```

Comment:
What's the expected behavior here if `offset` and `size` are outside the allocated range? Will this return nullptr? For instance, if the range is only `10` and I specify an offset of `5` and a `size` of `10`, will this return nullptr?

Reply:
Exactly, `ptr` will be nullptr and JSG_REQUIRE will bail out.